src/Nethermind/Nethermind.Consensus/GC/GCScheduler.cs

```diff
@@ -0,0 +1,192 @@
+    /// </summary>
+    /// <param name="queueCount">Number of items in the processing queue.</param>
+    public void SwitchOnBackgroundGC(int queueCount)
+    {
+        if (_fireGC)
+        {
+            _countToGC--;
+            // Trigger GC if block limit reached and minimum time elapsed
+            if (_countToGC <= 0 && _stopwatch.Elapsed.TotalSeconds > MinSecondsBetweenForcedGC)
+            {
+                _fireGC = false;
+                _stopwatch.Reset();
+                // Ensure only one GC task runs at a time
+                if (_lastGcTask.IsCompleted)
+                {
+                    _lastGcTask = PerformFullGCAsync();
+                }
+            }
+        }
+
+        // Avoid setting the timer if there are items in the queue or timer is already set
+        if (queueCount > 0 || _gcTimerSet)
+        {
+            // Don't switch on if still processing blocks
+            return;
+        }
+        _gcTimerSet = true;
+        // Schedule GC every 2 minutes when idle
+        _gcTimer.Change(TimeSpan.FromMinutes(2), TimeSpan.FromMinutes(2));
+    }
+
+    /// <summary>
+    /// Deactivates background garbage collection when the processing queue is active.
+    /// </summary>
+    /// <param name="queueCount">Number of items in the processing queue.</param>
+    public void SwitchOffBackgroundGC(int queueCount)
+    {
+        if (!_fireGC && queueCount > BlocksBacklogTriggeringManualGC)
+        {
+            // Long chains in archive sync don't force GC and don't call MallocTrim
+            // Start manual GC due to high backlog
+            _fireGC = true;
+            _countToGC = MaxBlocksWithoutGC;
+            _stopwatch.Restart();
+        }
+        else if (queueCount == 0)
+        {
+            // Stop manual GC when queue is empty
+            _fireGC = false;
+        }
+
+        if (!_gcTimerSet)
+        {
+            return;
+        }
+        // Stop the GC timer as the system is no longer idle
+        _gcTimerSet = false;
+        _gcTimer.Change(Timeout.Infinite, Timeout.Infinite);
+    }
+
+    /// <summary>
+    /// Performs a full garbage collection asynchronously to prevent blocking the main thread.
+    /// </summary>
+    private async Task PerformFullGCAsync()
+    {
+        // Switch to a background thread
+        await Task.Yield();
+
+        PerformFullGC();
+    }
+
+    /// <summary>
+    /// Marks that a garbage collection is in progress; or one shouldn't happen.
+    /// </summary>
+    /// <returns>True if marking succeeded; false if a GC is already in progress.</returns>
+    public static bool MarkGCPaused()
+    {
+        return Interlocked.CompareExchange(ref _isPerformingGC, 1, 0) == 0;
+    }
+
+    /// <summary>
+    /// Marks that garbage collection has finished.
+    /// </summary>
+    public static void MarkGCResumed()
+    {
+        Volatile.Write(ref _isPerformingGC, 0);
+    }
```

Comment:
So instead `_isPerformingGC` we should have `_canPerformGC` with `0` being `true`?

Reply:
.NET 9 allows enums to be used with `CompareExchange` so we can clean up some of these unhelpful 0, 1

Renamed and added some constants